src/simple_controller_impl/src/main.rs

```diff
@@ -0,0 +1,204 @@
+// Nightly clippy (0.1.64) considers Drop a side effect, see https://github.com/rust-lang/rust-clippy/issues/9608
+#![allow(clippy::unnecessary_lazy_evaluations)]
+
+use anyhow::Result;
+use futures::StreamExt;
+use k8s_openapi::api::core::v1::ConfigMap;
+use kube::{
+    api::{Api, ListParams, ObjectMeta, PostParams},
+    runtime::controller::{Action, Controller},
+    Client, CustomResource,
+};
+use kube_client;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::time::Duration;
+use tracing::*;
+
+#[derive(Debug, Error)]
+enum Error {
+    #[error("MissingObjectKey: {0}")]
+    MissingObjectKey(&'static str),
+}
+
+#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[kube(group = "anvil.dev", version = "v1", kind = "SimpleCR")]
+#[kube(shortname = "cr", namespaced)]
+struct SimpleCRSpec {}
+
+// The local state used in each reconcile core invocation
+// It is very simple and only has a pc
+struct SimpleReconcileState {
+    pub reconcile_pc: i32,
+}
+
+// This is the key of the custom resource object that the controller is reconciling for
+// TODO: add the type of the resource (i.e., kind) into it
+#[derive(Clone)]
+struct ResourceKey {
+    name: String,
+    namespace: String,
+    // kind: String,
+}
+
+// This is the API request that reconcile_core returns to reconcile
+// I did not find a way to encode the type into the enum
+// so each request operation is hardcoded to one type
+// This is a big limitation and we should work on it
+enum APIRequest {
```

Comment:
Could you elaborate on this?

Reply:
I want the APIRequest enum to be not hardcoded to any resource type and to take a generic type T:
```rust
enum APIRequest<T> {
    GetReq(...),
    CreateReq(...),
    DeleteReq(...),
    ...
}
```
then I need to way to tell the reconcile function which type of resource to get/create/delete.

Reply:
Maybe there is a way around it. I will try again